chain/network/src/snapshot_hosts/tests.rs

```diff
@@ -200,3 +202,143 @@ async fn test_lru_eviction() {
     // check that the oldest data was evicted
     assert_eq!([&info1, &info2].as_set(), cache.get_hosts().iter().collect::<HashSet<_>>());
 }
+
+// In each test, we will have a list of these, where they will indicate the function we
+// should call on the SnapshotHostsCache, and in the case of CallSelect, also the return value
+// we should expect to get, in this case expressed as the rank of the returned PeerId by priority score
+#[derive(Debug)]
+enum SelectPeerAction {
+    CallSelect(Option<usize>),
+    InsertHosts(&'static [usize]),
+    PartReceived,
+}
+
+struct SelectPeerTest {
+    num_peers: usize,
+    actions: &'static [SelectPeerAction],
+}
+
+static SELECT_PEER_CASES: &[SelectPeerTest] = &[
+    SelectPeerTest {
+        num_peers: 2,
+        actions: &[
+            SelectPeerAction::CallSelect(None),
+            SelectPeerAction::InsertHosts(&[0, 1]),
+            SelectPeerAction::CallSelect(Some(0)),
+            SelectPeerAction::CallSelect(Some(1)),
+            SelectPeerAction::CallSelect(Some(0)),
+            SelectPeerAction::CallSelect(Some(1)),
+        ],
+    },
+    SelectPeerTest {
+        num_peers: 3,
+        actions: &[
+            SelectPeerAction::CallSelect(None),
+            SelectPeerAction::CallSelect(None),
+            SelectPeerAction::InsertHosts(&[1, 2]),
+            SelectPeerAction::CallSelect(Some(1)),
+            SelectPeerAction::CallSelect(Some(2)),
+            SelectPeerAction::CallSelect(Some(1)),
+            SelectPeerAction::InsertHosts(&[0]),
+            SelectPeerAction::CallSelect(Some(0)),
+            // Now 0 and 2 have been returned once and 1 has been returned twice.
```

Comment:
yeah that is a good point... maybe in a separate PR we should try to deal with this somehow. what do you think about this idea?

so instead of just a simple `times_returned` that we're storing and comparing by, what if we try to guess sort of a "likelihood this peer might respond with the part" that gets informed by the caller of `select_host` later informing the `SnapshotHostsCache` of what happened. If the part was returned and validated, that counts in that peer's favor. If it was returned and was invalid, that is very negative.

If the peer didn't respond, that maybe slightly counts against that peer, but it's not so negative because the node might be down for maintenance, etc, or for any number of reasons. If you try a peer a bunch of times and it works each time, and then all of a sudden it doesn't respond, then we should move that kind of further back in the queue because it's not super likely to respond if we try again in 10 seconds. But if we try again in 3 days there's probably a good chance it will respond, so maybe some kind of time-weighted non-response count or something.

Then these stats could be put in some kind of struct and used as a comparator that hopefully won't have the probelm you describe here